# Exploring Ebay Car Sales Data

Data from eBay Kleinanzeigen, a classifieds section of German eBay site
Dataset found [here](https://www.kaggle.com/orgesleka/used-cars-database/data)
    - data is clean - it has been modified to more closely resemble a scraped dataset
I will clean the data and analyse the used car listings using panda

In [97]:
import pandas as pd
import numpy as np
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')


In [98]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [99]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Autos dataset contains 20 columns, most of which are strings
Some columns have null values, but none have > 20% null values
Column names use camelcase instead of snakecase - cannot replace spaces with underscores

1) convert camelcase to snakecase and reword some of the column names to be more descriptive based on the data dictionary


In [100]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [101]:
new_cols = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.columns = new_cols
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

Column names changed

Now, explore data to see if more cleaning needed

2) Look for:
    - text columns where (almost) all the values are the same - not useful for analysis
    - examples of numeric data stored as text which can be cleaned and converted

In [102]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

columns with potentially mostly one value - 2 unique and 49999 freq:
    - seller
    - offer_type
    

In [103]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [104]:
autos.drop('seller', axis=1, inplace=True)

In [105]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [106]:
autos.drop('offer_type', axis=1, inplace=True)

new dataframe:

In [107]:
autos.describe(include='all')

date_crawled         name  price abtest vehicle_type  \
count                 50000        50000  50000  50000        44905   
unique                48213        38754   2357      2            8   
top     2016-03-19 17:36:18  Ford_Fiesta     $0   test    limousine   
freq                      3           78   1421  25756        12859   
mean                    NaN          NaN    NaN    NaN          NaN   
std                     NaN          NaN    NaN    NaN          NaN   
min                     NaN          NaN    NaN    NaN          NaN   
25%                     NaN          NaN    NaN    NaN          NaN   
50%                     NaN          NaN    NaN    NaN          NaN   
75%                     NaN          NaN    NaN    NaN          NaN   
max                     NaN          NaN    NaN    NaN          NaN   

        registration_year  gearbox      power_ps  model   odometer  \
count        50000.000000    47320  50000.000000  47242      50000   
unique                NaN        2           NaN    245         13   
top                   NaN  manuell           NaN   golf  150,000km   
freq                  NaN    36993           NaN   4024      32424   
mean          2005.073280      NaN    116.355920    NaN        NaN   
std            105.712813      NaN    209.216627    NaN        NaN   
min           1000.000000      NaN      0.000000    NaN        NaN   
25%           1999.000000      NaN     70.000000    NaN        NaN   
50%           2003.000000      NaN    105.000000    NaN        NaN   
75%           2008.000000      NaN    150.000000    NaN        NaN   
max           9999.000000      NaN  17700.000000    NaN        NaN   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [108]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

clean price and odometer columns because they have numeric data stored as text

In [109]:
autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)
autos.rename(columns={'price':'price_$'}, inplace=True)

In [110]:
autos['odometer']=autos['odometer'].str.replace('km', '').str.replace(',','').astype(int)
autos.rename(columns={'odometer':'odometer_km'}, inplace=True)

In [111]:
autos['price_$'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price_$, dtype: int64

In [112]:
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

analyze odometer and price columns to look for outliers

In [113]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

no max outliers for odometer

In [114]:
autos['odometer_km'].value_counts().sort_index()

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

no min outliers for odometer

In [115]:
autos['price_$'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64

In [116]:
autos['price_$'].value_counts().sort_index(ascending=False).head(50)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
145000      1
139997      1
137999      1
135000      1
130000      1
129000      1
128000      1
120000      2
119900      1
119500      1
116000      1
115991      1
115000      1
114400      1
109999      1
105000      2
104900      1
99900       2
99000       2
98500       1
Name: price_$, dtype: int64

some cars have abnormally high prices in the millions of dollars. These should be removed.

$350,000 is still high but I will remove all cars priced higher than this to start and check the description

In [117]:
autos[autos['price_$'] > 350000]

date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   

        price_$   abtest vehicle_type  registration_year    gearbox  power_ps  \
514      999999     test        kombi               2009    manuell       101   
2897   11111111     test    limousine               1973    manuell        48   
7814    1300000  control        coupe               1992        NaN         0   
11137  10000000  control        coupe               1960    manuell       368   
22947   1234566  control        kombi               1999  automatik       190   
24384  11111111     test          NaN               1995        NaN         0   
27371  12345678  control          NaN               2017        NaN        95   
37585    999990     test    limousine               1985    manuell       111   
39377  12345678  control          NaN               2018    manuell        95   
39705  99999999  control    limousine               1999  automatik       224   
42221  27322222  control    limousine               2014    manuell       163   
43049    999999     test          bus               1981    manuell        70   
47598  12345678  control    limousine               2001    manuell       101   
47634   3890000     test        coupe               2006        NaN       799   

             model  odometer_km  registration_month fuel_type           brand  \
514          focus       125000                   4    benzin            ford   
2897        escort        50000                   3    benzin            ford   
7814           NaN        50000                  12       NaN  sonstige_autos   
11137          NaN       100000                   1    benzin  sonstige_autos   
22947          NaN       150000                   2    diesel             bmw   
24384          NaN       150000                   0       NaN      volkswagen   
27371        punto       150000                   0       NaN            fiat   
37585        jetta       150000                  12    benzin      volkswagen   
39377          v40       150000                   6       NaN           volvo   
39705     s_klasse       150000                   9    benzin   mercedes_benz   
42221           c4        40000                   2    diesel         citroen   
43049  transporter       150000                   1    benzin      volkswagen   
47598       vectra       150000                   3    benzin            opel   
47634          NaN         5000                   7       NaN  sonstige_autos   

      unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
514                nein  2016-03-17 00:00:00               0        12205   
2897               nein  2016-03-12 00:00:00               0        94469   
7814               nein  2016-04-04 00:00:00               0        60598   
11137              nein  2016-03-29 00:00:00              

In [118]:
autos[~(autos['price_$'] > 350000)].shape[0]

49986

In [119]:
autos = autos[~(autos['price_$'] > 350000)]

In [120]:
autos.shape[0]

49986

In [121]:
autos['price_$'].describe()

count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price_$, dtype: float64

In [122]:
autos['price_$'].value_counts().sort_index(ascending=False).head(50)

350000    1
345000    1
299000    1
295000    1
265000    1
259000    1
250000    1
220000    1
198000    1
197000    1
194000    1
190000    1
180000    1
175000    1
169999    1
169000    1
163991    1
163500    1
155000    1
151990    1
145000    1
139997    1
137999    1
135000    1
130000    1
129000    1
128000    1
120000    2
119900    1
119500    1
116000    1
115991    1
115000    1
114400    1
109999    1
105000    2
104900    1
99900     2
99000     2
98500     1
94999     1
93911     1
93000     2
89900     1
89000     1
88900     1
86500     1
85000     1
84997     1
84000     1
Name: price_$, dtype: int64

In [123]:
autos[autos['price_$'] > 70000].shape[0]

82

82 cars over $70,000, will check to see if worth deleting 

In [124]:
autos_check = autos[~(autos['price_$'] > 70000)]

In [125]:
autos_check.shape[0]

49904

In [126]:
autos_check['price_$'].describe()

count    49904.000000
mean      5525.635941
std       7118.136470
min          0.000000
25%       1100.000000
50%       2900.000000
75%       7200.000000
max      70000.000000
Name: price_$, dtype: float64

barely changed the mean, changed std deviation by ~2,000. Will not remove cars over 1k

In [127]:
autos['price_$'].value_counts().sort_index().head(50)

0      1421
1       156
2         3
3         1
5         2
8         1
9         1
10        7
11        2
12        3
13        2
14        1
15        2
17        3
18        1
20        4
25        5
29        1
30        7
35        1
40        6
45        4
47        1
49        4
50       49
55        2
59        1
60        9
65        5
66        1
70       10
75        5
79        1
80       15
89        1
90        5
99       19
100     134
110       3
111       2
115       2
117       1
120      39
122       1
125       8
129       1
130      15
135       1
139       1
140       9
Name: price_$, dtype: int64

will remove cars at 0 but keep cars at price >= 1

In [128]:
autos_check = autos[autos['price_$'] > 0]

In [129]:
autos_check['price_$'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_$, dtype: float64

In [130]:
autos = autos[autos['price_$'] > 0]

removed free cars from dataframe

In [131]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
30     2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
...                    ...                                                ...   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

5 columns referring to dates
from data dictionary:
    - `date_crawled`: added by the crawler
    - `last_seen`: added by the crawler
    - `ad_created`: from the website
    - `registration_month`: from the website
    - `registration_year`: from the website

date_crawled, last_seen, and ad_created all identified as string values - need to convert to numerical representation
resistration month and year represented as numeric values

string values all in timestamp form. We care about the date and not the time
    - instead of using datetime we could easily estract the date using the first 10 characters
    
date_crawled:

In [132]:
date_crawled = autos['date_crawled'].str[:10]

In [133]:
date_crawled[0:5]

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: date_crawled, dtype: object

In [134]:
date_crawled.value_counts(normalize=True, dropna=False).sort_index() * 100

2016-03-05    2.532688
2016-03-06    1.404304
2016-03-07    3.601359
2016-03-08    3.329558
2016-03-09    3.308967
2016-03-10    3.218367
2016-03-11    3.257490
2016-03-12    3.691959
2016-03-13    1.566972
2016-03-14    3.654896
2016-03-15    3.428395
2016-03-16    2.960980
2016-03-17    3.162772
2016-03-18    1.291053
2016-03-19    3.477813
2016-03-20    3.788737
2016-03-21    3.737259
2016-03-22    3.298672
2016-03-23    3.222485
2016-03-24    2.934212
2016-03-25    3.160712
2016-03-26    3.220426
2016-03-27    3.109235
2016-03-28    3.486050
2016-03-29    3.409863
2016-03-30    3.368681
2016-03-31    3.183363
2016-04-01    3.368681
2016-04-02    3.547823
2016-04-03    3.860805
2016-04-04    3.648718
2016-04-05    1.309585
2016-04-06    0.317101
2016-04-07    0.140019
Name: date_crawled, dtype: float64

In [135]:
date_crawled.describe()

count          48565
unique            34
top       2016-04-03
freq            1875
Name: date_crawled, dtype: object

In date_crawled, each date crawled approx. 3% of the total data. There are 34 total dates where data was crawled

last_seen:

In [136]:
last_seen = autos['last_seen'].str[:10]

In [137]:
last_seen.value_counts(normalize=True, dropna=False).sort_index() * 100

2016-03-05     0.107073
2016-03-06     0.432410
2016-03-07     0.539483
2016-03-08     0.741275
2016-03-09     0.959539
2016-03-10     1.066612
2016-03-11     1.237517
2016-03-12     2.378256
2016-03-13     0.889529
2016-03-14     1.260167
2016-03-15     1.587563
2016-03-16     1.645218
2016-03-17     2.808607
2016-03-18     0.735097
2016-03-19     1.583445
2016-03-20     2.065273
2016-03-21     2.063214
2016-03-22     2.137342
2016-03-23     1.853186
2016-03-24     1.976732
2016-03-25     1.921137
2016-03-26     1.680222
2016-03-27     1.564913
2016-03-28     2.085864
2016-03-29     2.234119
2016-03-30     2.477093
2016-03-31     2.378256
2016-04-01     2.279419
2016-04-02     2.491506
2016-04-03     2.520334
2016-04-04     2.448265
2016-04-05    12.476063
2016-04-06    22.180583
2016-04-07    13.194688
Name: last_seen, dtype: float64

In [138]:
last_seen.describe()

count          48565
unique            34
top       2016-04-06
freq           10772
Name: last_seen, dtype: object

last seen dates are the same as dates crawled because they both were added by the crawler
the amount of cars last seen seem to increase as time goes on, which makes sense
the top three dates for last seen were in the past last three days of crawling

ad_created:

In [139]:
ad_created = autos['ad_created'].str[:10]

In [140]:
ad_created.value_counts(normalize=True, dropna=False).sort_index() * 100

2015-06-11    0.002059
2015-08-10    0.002059
2015-09-09    0.002059
2015-11-10    0.002059
2015-12-05    0.002059
2015-12-30    0.002059
2016-01-03    0.002059
2016-01-07    0.002059
2016-01-10    0.004118
2016-01-13    0.002059
2016-01-14    0.002059
2016-01-16    0.002059
2016-01-22    0.002059
2016-01-27    0.006177
2016-01-29    0.002059
2016-02-01    0.002059
2016-02-02    0.004118
2016-02-05    0.004118
2016-02-07    0.002059
2016-02-08    0.002059
2016-02-09    0.002059
2016-02-11    0.002059
2016-02-12    0.004118
2016-02-14    0.004118
2016-02-16    0.002059
2016-02-17    0.002059
2016-02-18    0.004118
2016-02-19    0.006177
2016-02-20    0.004118
2016-02-21    0.006177
                ...   
2016-03-09    3.315145
2016-03-10    3.189540
2016-03-11    3.290435
2016-03-12    3.675486
2016-03-13    1.700813
2016-03-14    3.518995
2016-03-15    3.401627
2016-03-16    3.012458
2016-03-17    3.127767
2016-03-18    1.359003
2016-03-19    3.368681
2016-03-20    3.794914
2016-03-21 

In [141]:
ad_created.describe()

count          48565
unique            76
top       2016-04-03
freq            1887
Name: ad_created, dtype: object

ad_created contains 76 dates. it seems that more ads were created later in the year, and barely any were created earlier in the year

In [142]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [143]:
autos['registration_year'].value_counts().sort_index()

1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      23
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64

registration year has outliers that are impossible. I will remove them

In [144]:
autos = autos[~((autos['registration_year'] < 1910) | (autos['registration_year'] > 2020))]

In [147]:
autos['registration_year'].value_counts(normalize=True).sort_index() * 100

1910    0.010300
1927    0.002060
1929    0.002060
1931    0.002060
1934    0.004120
1937    0.008240
1938    0.002060
1939    0.002060
1941    0.004120
1943    0.002060
1948    0.002060
1950    0.006180
1951    0.004120
1952    0.002060
1953    0.002060
1954    0.004120
1955    0.004120
1956    0.008240
1957    0.004120
1958    0.008240
1959    0.012360
1960    0.047379
1961    0.012360
1962    0.008240
1963    0.016480
1964    0.024719
1965    0.035019
1966    0.045319
1967    0.053559
1968    0.053559
          ...   
1990    0.714801
1991    0.698321
1992    0.762179
1993    0.875476
1994    1.295705
1995    2.527552
1996    2.828304
1997    4.018951
1998    4.867649
1999    5.967659
2000    6.501184
2001    5.430013
2002    5.121022
2003    5.559790
2004    5.568030
2005    6.047997
2006    5.500051
2007    4.690493
2008    4.562777
2009    4.294984
2010    3.273252
2011    3.343290
2012    2.698527
2013    1.654135
2014    1.365743
2015    0.807498
2016    2.513132
2017    2.8674

In [146]:
autos['registration_year'].describe()

count    48545.000000
mean      2003.461510
std          7.566196
min       1910.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2019.000000
Name: registration_year, dtype: float64

average car registration year is around may 2003, +- 7.5 years

To analyse the brand column, I will use aggregation:

In [172]:
brand_value_counts = autos['brand'].value_counts(normalize=True).sort_index() * 100
brand_value_counts

alfa_romeo         0.661242
audi               8.585848
bmw               10.864147
chevrolet          0.566485
chrysler           0.348131
citroen            1.411062
dacia              0.265733
daewoo             0.156556
daihatsu           0.251313
fiat               2.599650
ford               6.966732
honda              0.799258
hyundai            0.994953
jaguar             0.152436
jeep               0.220414
kia                0.710681
lada               0.059738
lancia             0.113297
land_rover         0.203934
mazda              1.522299
mercedes_benz      9.578741
mini               0.861057
mitsubishi         0.817798
nissan             1.526419
opel              10.864147
peugeot            2.945720
porsche            0.591204
renault            4.789371
rover              0.133896
saab               0.162736
seat               1.893089
skoda              1.606757
smart              1.429601
sonstige_autos     0.964054
subaru             0.210114
suzuki             0

In [174]:
brand_index = autos['brand'].value_counts().sort_index().index
brand_index

Index(['alfa_romeo', 'audi', 'bmw', 'chevrolet', 'chrysler', 'citroen',
       'dacia', 'daewoo', 'daihatsu', 'fiat', 'ford', 'honda', 'hyundai',
       'jaguar', 'jeep', 'kia', 'lada', 'lancia', 'land_rover', 'mazda',
       'mercedes_benz', 'mini', 'mitsubishi', 'nissan', 'opel', 'peugeot',
       'porsche', 'renault', 'rover', 'saab', 'seat', 'skoda', 'smart',
       'sonstige_autos', 'subaru', 'suzuki', 'toyota', 'trabant', 'volkswagen',
       'volvo'],
      dtype='object')

will aggregate only the brands that make up > 1% of the total

In [177]:
agg_brands = []
for brand in brand_index:
    if brand_value_counts[brand] > 1:
        agg_brands.append(brand)
print(agg_brands)

['audi', 'bmw', 'citroen', 'fiat', 'ford', 'mazda', 'mercedes_benz', 'nissan', 'opel', 'peugeot', 'renault', 'seat', 'skoda', 'smart', 'toyota', 'volkswagen']


In [196]:
brands_dict = {}
for brand in agg_brands:
    brand_prices = autos[autos['brand'] == brand]['price_$']
    brands_dict[brand] = round(brand_prices.mean(),2)
    
print(brands_dict)

{'renault': 2431.2, 'audi': 9212.93, 'bmw': 8261.38, 'fiat': 2793.87, 'seat': 4320.17, 'smart': 3518.1, 'toyota': 5148.0, 'opel': 2941.47, 'mazda': 4059.06, 'ford': 3728.41, 'nissan': 4669.39, 'volkswagen': 5333.2, 'peugeot': 3065.61, 'citroen': 3756.07, 'mercedes_benz': 8526.62, 'skoda': 6353.54}


In [220]:
avg_price = pd.Series(brands_dict)
avg_price.sort_values(ascending = False)

audi             9212.93
mercedes_benz    8526.62
bmw              8261.38
skoda            6353.54
volkswagen       5333.20
toyota           5148.00
nissan           4669.39
seat             4320.17
mazda            4059.06
citroen          3756.07
ford             3728.41
smart            3518.10
peugeot          3065.61
opel             2941.47
fiat             2793.87
renault          2431.20
dtype: float64

Of the common brands on the site, Audi is the most expensive, with Mercedes Benz and BMW following close behind. Renault, Fiat, and Opel are the three least expensive brands

Now I will compare these average brand prices to their average milages to see if there is a correlation between milage and price

In [221]:
odometer_dict = {}
for brand in agg_brands:
    brand_odometer = autos[autos['brand'] == brand]['odometer_km']
    odometer_dict[brand] = round(brand_odometer.mean(),2)
    
print(odometer_dict)

{'renault': 128062.37, 'audi': 129492.56, 'bmw': 132682.97, 'fiat': 117567.35, 'seat': 121768.23, 'smart': 100511.53, 'toyota': 116219.31, 'opel': 129452.03, 'mazda': 124871.45, 'ford': 124349.5, 'nissan': 118711.2, 'volkswagen': 128955.57, 'peugeot': 127356.64, 'citroen': 120160.58, 'mercedes_benz': 130848.39, 'skoda': 110993.59}


In [227]:
avg_miles = pd.Series(odometer_dict)
price_miles = pd.DataFrame(avg_price, columns = ['mean_price'])
price_miles['mean_miles'] = avg_miles
price_miles

mean_price  mean_miles
audi              9212.93   129492.56
bmw               8261.38   132682.97
citroen           3756.07   120160.58
fiat              2793.87   117567.35
ford              3728.41   124349.50
mazda             4059.06   124871.45
mercedes_benz     8526.62   130848.39
nissan            4669.39   118711.20
opel              2941.47   129452.03
peugeot           3065.61   127356.64
renault           2431.20   128062.37
seat              4320.17   121768.23
skoda             6353.54   110993.59
smart             3518.10   100511.53
toyota            5148.00   116219.31
volkswagen        5333.20   128955.57

It appears that milage doesn't affect price by a whole lot. The most expensive brands have some of the highest miles in the DataFrame, and Smart car with the lowest miles is has one of the lowest average prices.

This should have been expected. People with expensive car brands with low miles tend to sell on places other than eBay, so the expensive brands will naturally have high milage. 